csrank/metrics.py

```diff
@@ -108,16 +108,59 @@ def zero_one_rank_loss_for_scores_ties(y_true, s_pred):
 
 
 def make_ndcg_at_k_loss(k=5):
+    r"""Computes the Normalized Discounted Cumulative Gain
+
+    The Discounted Cumulative Gain is the sum of the document's relevancies,
+    logarithmically discounted by their rank. That means the DCG is higher when
+    the more relevant documents are highly ranked, lower otherwise.
+
+    Concretely: TODO document that we actually use the 2**rel variant
+
+    .. math::
+        \mathrm{DCG}_p = \sum_{i = 1}^p \frac{\mathit{rel}_i}{\log_2(i + 1)}
+
+    Where :math:`\mathit{rel}_i` is the relevance of the document that is
+    ranked at :math:`i`.
+
+    To make the DCG comparable across different ranking (particularly rankings
+    of different length), it is normalized by the ideal DCG. The resulting nDCG
+    can be described as
+
+    .. math::
+        \mathrm{nDCG}_p = \frac{\mathrm{DCG}_p}{\mathrm{IDCG}_p}
+
+    with
+
+    .. math::
+        \mathrm{IDCG}_p
+        = \sum_{i = 1}^{\lvert \mathit{REL}_p \rvert} \frac{mathit{rel}}{\log_2(i + 1)}
+
+    where :math:`\mathit{REL}_p` is the list of relevant documents and
+    `\mathit{rel_i}` are the document relevancies in decreasing order.
+
+    It follows that the nDCG is always a value in :math:`(0, 1]`, with
+    :math:`1` being the best value.
+
+    Parameters
+    ----------
+    k: int
+        The length of the ranking.
+    """
     def ndcg(y_true, y_pred):
         y_true, y_pred = tensorify(y_true), tensorify(y_pred)
-        n_objects = K.cast(K.int_shape(y_pred)[1], 'float32')
-        relevance = K.pow(2., ((n_objects - y_true) * 60) / n_objects) - 1.
-        relevance_pred = K.pow(2., ((n_objects - y_pred) * 60) / n_objects) - 1.
+        max_rank = K.cast(K.int_shape(y_pred)[1], 'float32')
+        # define the relevance as 2**a
+        relevance = n_objects - y_true # ranking on place 1 has the highest relevance
+        # TODO Where does the 60 come from?
```

Comment:
@prithagupta you added the `*60` in 7bc279d4607cf56ae9c4d9eebd5a582117ae0989. What is the idea behind that?

Reply:
@prithagupta ping :)

Reply:
It's been long, time since I did this, but I had to do this to avoid the nan problems. 
NDCG requires the relevance score.
For prediction, we get the predicted relevance score.
For ground truth, we just have rankings in the real/synthetic datasets.
And there is no standard way to convert rankings to relevance.
For letor-ranking dataset the ground truth only consists of relevance score 0,1,2. While for a long ranking such as 1200.
So, the exponential gives us a NAN.

Reply:
I did a quick calculation and I see no reason not to normalize the inverse ranks (relevance) to [0, 1].
Here is a plot of what that could look like:
![dcg01](https://user-images.githubusercontent.com/564473/76944230-f1734000-6900-11ea-9a72-141ae0a1d14b.png)

Example: If we have 3 objects A, B and C and the ranking A ≻ C ≻ B, then
```python
ranking = [0, 2, 1]
inverse_ranking = 2 - ranking = [2, 0, 1]  # or relevance
rel = inverse_ranking / 2 = [1, 0, 0.5]
```

Reply:
I'm not sure what exactly your intention is with that plot. Normalizing certainly changes the ndcg, but it shouldn't change relative comparisons and since our rank<->relevancy conversion is arbitrary anyway I don't see a reason not to do it either.

Reply:
Yes, the conversion is completely arbitrary. But if we always normalize to [0, 1], it is at least comparable for different ranking sizes.